src/swarm-lang/Swarm/Language/Parser/Term.hs

```diff
@@ -92,11 +93,20 @@ parseTermAtom2 =
           <*> optional (symbol ":" *> parsePolytype)
           <*> (symbol "=" *> parseTerm)
           <*> (reserved "in" *> parseTerm)
-        <|> sLet LSDef
-          <$> (reserved "def" *> locTmVar)
-          <*> optional (symbol ":" *> parsePolytype)
-          <*> (symbol "=" *> parseTerm <* reserved "end")
-          <*> (optional (symbol ";") *> (parseTerm <|> (eof $> sNoop)))
+        -- <|> sLet LSDef
+        --   <$> (reserved "def" *> locTmVar)
+        --   <*> optional (symbol ":" *> parsePolytype)
+        --   <*> (symbol "=" *> parseTerm <* reserved "end")
+        --   <*> (optional (symbol ";") *> (parseTerm <|> (eof $> sNoop)))
+        <|> do
+          reserved "def"
+          locVar@(LV _srcLoc nameText) <- locTmVar
+          mTy <- optional (symbol ":" *> parsePolytype)
+          _ <- symbol "="
+          body <- parseTerm
+          reserved "end" <?> ("missing 'end' keyword for definition of '" <> T.unpack nameText <> "'")
```

Comment:
I had not thought about doing this in the parser!  Does it work?  I worry that the `parseTerm` right before `reserved "end"` will be quite happy to keep going, including the subsequent `def`.  I think it won't realize anything is wrong until it gets all the way to the end and realizes the original `def` was never closed by `end`...

Reply:
Kinda? The message ends up as one of the possible expected alternatives so you see:

"4:1:\n  |\n4 | <empty line>\n  | ^\nunexpected end of input\nexpecting \"!=\", \"&&\", \"()\", \"++\", \"<=\", \"==\", \">=\", \"def\", \"false\", \"let\", \"require\", \"requirements\", \"true\", \"tydef\", \"||\", '\"', '$', '(', '*', '+', '-', '.', '/', ':', ';', '<', '>', '[', '\\', '^', **'end' keyword for definition of 'a'**, '{', built-in user function, direction constant, integer literal, or variable name\n"

Which isn't a very clear message...  Maybe I'm taking the wrong approach.  Do you have any suggestions about where would be a better place to implement the check?

Reply:
Oh, I see!  That makes sense.  Given that it *is* now possible to nest definitions, it might be the best we can do.